examples/todomvc/src/todo.rs

```diff
@@ -0,0 +1,229 @@
+
+    /// Begin editing the TODO item with the given id.
+    fn begin_editing(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Update the edits for the TODO with the given id.
+    fn update_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize, edits: String);
+
+    /// Finish editing the TODO with the given id.
+    fn finish_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+
+    /// Cancel editing the TODO with the given id.
+    fn cancel_edits(root: &mut dyn RootRender, vdom: VdomWeak, id: usize);
+}
+
+impl<C> Todo<C> {
+    /// Construct a new `Todo` with the given identifier and title.
+    pub fn new<S: Into<String>>(id: usize, title: S) -> Self {
+        let title = title.into();
+        let completed = false;
+        let edits = None;
+        Todo {
+            id,
+            title,
+            completed,
+            edits,
+            _controller: PhantomData,
+        }
+    }
+
+    /// Set this TODO item's id.
+    pub fn set_id(&mut self, id: usize) {
+        self.id = id;
+    }
+
+    /// Is this `Todo` complete?
+    pub fn is_complete(&self) -> bool {
+        self.completed
+    }
+
+    /// Mark the `Todo` as complete or not.
+    pub fn set_complete(&mut self, to: bool) {
+        self.completed = to;
+    }
+
+    /// Get this TODO's title.
+    pub fn title(&self) -> &str {
+        &self.title
+    }
+
+    /// Set this TODO item's title.
+    pub fn set_title<S: Into<String>>(&mut self, title: S) {
+        self.title = title.into();
+    }
+
+    /// Set the edits for this TODO.
+    pub fn set_edits<S: Into<String>>(&mut self, edits: Option<S>) {
+        self.edits = edits.map(Into::into);
+    }
+
+    /// Take this TODO's edits, leaving `None` in their place.
+    pub fn take_edits(&mut self) -> Option<String> {
+        self.edits.take()
+    }
+}
+
+impl<C: TodoActions> Render for Todo<C> {
+    fn render<'a, 'bump>(&'a self, bump: &'bump Bump) -> dodrio::Node<'bump>
+    where
+        'a: 'bump,
+    {
+        use dodrio::bumpalo::{self, collections::String};
+
+        let mut class = String::new_in(bump);
+        if self.completed {
+            class.push_str("completed ");
+        }
+        if self.edits.is_some() {
+            class.push_str("editing");
+        }
+
+        let id = self.id;
+        let title = self.edits.as_ref().unwrap_or(&self.title);
+        let elem_id = bumpalo::format!(in bump, "todo-{}", id);
+
+        let mut input_attrs = bumpalo::vec![
+            in bump;
+            Attribute {
+                name: "class",
+                value: "toggle",
+            },
+            Attribute {
+                name: "type",
+                value: "checkbox",
+            }
+        ];
+        if self.completed {
+            input_attrs.push(Attribute {
+                name: "checked",
+                value: "",
+            });
+        }
+        let input_attrs = input_attrs.into_bump_slice();
+
+        Node::element(
```

Comment:
Heh we should definitely clean this up a bit at some point :)

(the helper methods you were mentioning would be perfect)

(well, that plus maybe a builder to add attributes)